### ISSUE TITLE
The plugin treats every OPTIONS request as a preflight request

### ISSUE THREAD

**Issue by TqrHsn** [fastify/fastify-cors#88]
#13  🐛 Bug Report

For an OPTIONS request to be a valid preflight request, it should have following Headers present in the request:
- Origin
- Access-Control-Request-Method
- Access-Control-Request-Headers
Otherwise the request should be treated as a normal OPTIONS request and forwarded to route

## To Reproduce
// package.json
```
{
  "main": "index.js",
  "scripts": {
    "start": "node index.js"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "fastify": "^3.4.1",
    "fastify-cors": "4.1.0"
  }
}
```

// index.js
```
const fastify = require('fastify')()

fastify.register(require('fastify-cors'), { 
    origin: "*",
    optionsSuccessStatus: 200,
    methods: "GET,HEAD,PUT,PATCH,POST,DELETE,OPTIONS",
})

fastify.get('/', (req, reply) => {
  reply.send({ hello: 'world' })
})

fastify.listen(3000)
```

// requests
1- With CORS headers
```
curl --location --request OPTIONS 'http://localhost:3000/' \
--header 'Origin: abc.com' \
--header 'Access-Control-Request-Method: POST'
```
Response: 200

2- Without CORS headers
```
curl --location --request OPTIONS 'http://localhost:3001/'
```
Resposne: 200

## Expected behavior

If the request is not a valid preflight request (does not contain CORS preflight request mandatory headers), it should be forwarded to route

## Your Environment

- *node version*: 10
- *fastify version*: >=2.0.0
- *os*: Windows

**Comment by mcollina** [comment]
Can you please include a full example to reproduce?

**Comment by TqrHsn** [comment]
@mcollina I have just added an example in the original issue

**Comment by mcollina** [comment]
What do you mean by "forwarded"? Should it include a `Location` header?

I'm sorry but I could not find any reference in the standard that specified this, could you send a link?

**Comment by TqrHsn** [comment]
Sorry, by "forwarded" I meant to call next() so that the request is passed to handle by middlewares etc.
Following is the link defining constraints for a preflight request:
[https://www.w3.org/TR/2020/SPSD-cors-20200602/#preflight-request](https://www.w3.org/TR/2020/SPSD-cors-20200602/#preflight-request)

**Comment by TqrHsn** [comment]
Another reference: [https://fetch.spec.whatwg.org/#http-requests](https://fetch.spec.whatwg.org/#http-requests)

**Comment by mcollina** [comment]
What should be the response if those headers are not present? That's what I am missing and what I do not see covered in the standard.

**Comment by TqrHsn** [comment]
The request should not be treated as a CORS preflight request and hence not handled by the plugin rather it should just call next()

**Comment by TqrHsn** [comment]
![image](https://user-images.githubusercontent.com/20878328/96275505-3f613b00-0feb-11eb-8762-4be3182e975f.png)

This snippet is from "CORS in Action" book.

**Comment by mcollina** [comment]
In fastify there is no "next" to call. Should it be a 404 then?

**Comment by TqrHsn** [comment]
Talking about [this](https://github.com/fastify/fastify-cors/blob/bfb2cf63ace5253ec599d46f7e7eac8ba0edc4ec/index.js#L39) next.

**Comment by mcollina** [comment]
I truly don't understand what HTTP status code and response you are proposing to send to the client. Calling `next` there will still return an empty 200.

**Comment by TqrHsn** [comment]
Calling next would forward the request to route. Then it would be up to the route what ever they return.

**Comment by jsumners** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by TqrHsn** [comment]
I can try on the weekend may be.

**Comment by mcollina** [comment]
@codyzu is this done?

**Comment by codyzu** [comment]
@mcollina I'm waiting for #103 to be merged and a new release to be made. v5.0.0 of fastify-cors introduced, and enabled by default, the `strictPreflight` option, but can potentially create conflicts if the preflight requests are rejected by an authentication plugin. #103 should be the perfect balance.

**Comment by codyzu** [comment]
@TqrHsn this should now be resolved with release 5.1.0 of fastify-cors. The `strictPreflight` option has been added and defaults to true. Options requests missing the Origin or Access-Control-Request-Method header will reply with 400 errors.

If your intention is to add OPTIONS routes for some purpose other than cors preflighting, set `strictPreflight` to false, `preflightContinue` to true, and handle the responses yourself. This is a highly unusual use case but should be possible.

@mcollina for me this can now be closed
